data_model/src/account.rs

```diff
@@ -104,17 +96,13 @@ pub mod model {
     pub struct Account {
         /// An Identification of the [`Account`].
         pub id: AccountId,
-        /// Assets in this [`Account`].
-        pub assets: AssetsMap,
```

Comment:
why would you move assets out of the account?

Reply:
To store every asset next to each other, we have different queries that iterate over assets inside particular domain.

Also in case of asset definition removal we iterate through every domain, every account inside it. It much effective to store them in one place.

Reply:
ok, I get it, you want to make `Account` more shallow. I do not agree with this necessarily.
The reason for removing `Account::roles` was because it can be expressed through a "special" asset given to the account. This way there would be no difference in API between public vs private blockchains because it would be expressed through the common concept. This is what I would have preferred

Reply:
> The reason for removing Account::roles was because it can be expressed through a "special" asset given to the account. This way there would be no difference in API between public vs private blockchains because it would be expressed through the common concept. This is what I would have preferred

But we still have `PermissionToken` in the `data_model` (through it's content could be arbitrary), do you think it should be removed too?

How usable it will to the user? Because users will have to query assets, write to assets instead or roles.

Reply:
> How usable it will to the user? Because users will have to query assets, write to assets instead or roles.

I don't see why necessarily. We can still keep the specialized API for dealing with roles (with `Grant` and `Revoke`). 
I still think `Account::roles` should be an asset of an account

> But we still have `PermissionToken` in the `data_model` (through it's content could be arbitrary), do you think it should be removed too?

I don't understand the question. If you're asking me whether I would prefer this to be an asset of account, then yes. Why? Because when a user queries `FindAccountById` it gets all the info about an account. 

Is cloning accounts something that we do extensively? I don't think so, and I don't think making accounts more shallow will bring performance benefits

Reply:
the way I thought this would work is that `Roles` asset definition would be created by the Executor. If the blockchain is private, i.e. will deal with the concept of roles, otherwise not. And `Grant`/`Revoke` would delegate to mint/burn `Roles` asset

Reply:
the way I thought this would work is that `Roles` asset definition would be created by the Executor. If the blockchain is private, i.e. will deal with the concept of roles, otherwise not. And `Grant`/`Revoke` would delegate to mint/burn `Roles` asset

Reply:
I wouldn't mind if we used `Mint`/`Burn` terminology instead `Grant`/`Revoke`. IMO they are just overloads of the same operation. In fact, I think this is what we should do. We want Iroha to be a simple blockchain with a simple API

if we can make our ISA extendable (by providing a mechanism to define custom instructions in the Executor). We would be able to add `Grant`/`Revoke` as custom instructions

Reply:
> Is cloning accounts something that we do extensively? I don't think so, and I don't think making accounts more shallow will bring performance benefits

Currently we do quite a few wsv clones which result in cloning of all accounts, also quires like `FindAllAccounts` result in account cloning.

> and I don't think making accounts more shallow will bring performance benefits

It bring us benefits in particular context such as when we need to iterate over assets (not necessary for single account). Because if's more effective to store entities next to each other without pointer indirection. 

> I don't understand the question. If you're asking me whether I would prefer this to be an asset of account, then yes. Why? Because when a user queries FindAccountById it gets all the info about an account.

Ok, so our plans is to remove `PermissionToken` and `Role` from the codebase, so it will be concepts defined on the level of executor, right? 

> We can still keep the specialized API for dealing with roles (with Grant and Revoke).

In such case having specialized API seems contradictory to me, because we might have some special API for possibly non-existent thing (if executor writer don't bother to use permissions and roles). Also we don't have a control how exactly this concepts will be implemented (maybe executor writer would decide to store permissions/roles) in account metadata not account assets.

Reply:
> We want Iroha to be a simple yet extendable blockchain with a simple API
> if we can make our ISA extendable (by providing a mechanism to define custom instructions in the Executor). We would be able to add Grant/Revoke as custom instructions

We probably should define what `Executor` is allowed to do. Extreme would case would be that iroha is responsible for storing data + consensus. And actual interpretation of data, execution of transactions will be executor responsibility.

Reply:
> Currently we do quite a few wsv clones which result in cloning of all accounts, also quires like FindAllAccounts result in account cloning.

1. if you store roles in the `wsv` cloning `wsv` will still clone all roles
2. yes, this is why we have a query `FindRolesByAccountId`. We shouldn't though, we should use filters

> Also we don't have a control how exactly this concepts will be implemented (maybe executor writer would decide to store permissions/roles) in account metadata not account assets.

IMO we should remove `Metadata` from an `Account`. It should also be just an asset of an account

Reply:
> We probably should define what Executor is allowed to do. Extreme would case would be that iroha is responsible for storing data + consensus. And actual interpretation of data, execution of transactions will be executor responsibility.

It could be seen as an interpreter that transforms higher level API into RISC instructions that operate directly on the blockchain

Reply:
@mversic, let's summarize this a bit:
1. let's try to remove `Role`s
2. do you think that `Asset`s should be returned to the `Account`? (we can gather other opinion's as well)

Reply:
To elaborate on the shallow account i see it as tables in `SQL` where we have foreign key to the entity in our case `Assets` already have `AccountId` which points to the `Account`.

Reply:
> @mversic, let's summarize this a bit:
> 1. let's try to remove `Role`s
> 2. do you think that `Asset`s should be returned to the `Account`? (we can gather other opinion's as well)

1. Yes, though I would prefer you make it into an asset right now, it's fine even if you make it part of wsv atm. Switch from wsv to asset will not be an API breaking change and we can do that later of it's too much work now
2. I would open an issue to discuss swallowing of accounts

Reply:
Created #3921, however scope is more broad.